### ISSUE TITLE
scope data returned as list of tuples

### ISSUE THREAD

**Issue by cycomanic** [python-ivi/python-ivi#49]
I was wondering what is the reason that the scopes return waveform data as a list of tuples instead of two arrays (one for time one for voltage) in `_measurement_fetch_waveform`? I can not see a case where one would prefer data in this format (e.g. for plotting one always wants separate arrays for x and y). 

Some background on why I'm asking this. I have an application where I want to fetch and display the waveform as close to real time as possible. The way the x and y waveforms are generated at the moment is of the form of

```python
data = [ ((i * xincr) + xzero, ((y - yoff) * ymult) + yzero) for i, y in enumerate(y_data)]
```

or similar. This is very slow. Compared to the following code:

```python
xdata = np.arange(len(y_data))*xincr + xzero
ydata = (y_data - yoff)*ymult + yzero
data = xdata, ydata
```

It is a factor 100-1000 times slower for 1e5 or 1e6 points (disregarding having to convert the output to an array for further processing, which adds a large additional processing time). 

I would really prefer the second method (or at least the ability to configure this somehow), I'm happy to create a pull request if this would be accepted.

**Comment by alexforencich** [comment]
Originally python-ivi did not have a numpy dependency and that seemed
like a good format, then I wanted to keep it consistent.  However,
you're right, I think it would be better to return the result in a
different format for higher performance.  There is a question on what
format would make the most sense, though.  I'm thinking the most ideal
format may be to create and return a trace object of some sort that
defers most of the conversion.  This way, you could deal with the raw
binary data if you really want to, or just the Y data and convert when
needed, or get both the Y data and the T data.  This can be especially
useful for large traces (many megapoints) as storing 2 floats or doubles
per point will take up a lot more space than storing a one or two bytes
per point plus some trace metadata. 

Thoughts?

Alex Forencich

**Comment by cycomanic** [comment]
I do agree with this approach, alternatively I would consider converting the Y data to integers (without converting to voltages), as I think most applications would want it in at least this format for further processing (most of my use cases are actually like this and I never use the T data).
I would also provide some convenience methods/functions that convert to the different formats. 

I just did a quick grep through the sources and the only trace class I found is the Multitrace class (which is not what you meant by trace object I assume). What do you think would be the best way of implementing this? Subclassing numpy array could be one way?

Jochen

**Comment by alexforencich** [comment]
What do you think about this approach?

https://github.com/alexforencich/python-ivi/commit/b94845c0ed5317bd0f5578d2547ef0c3b5d77f44

**Comment by cycomanic** [comment]
Yes, this works. As I said I was considering of subclassing ndarray and simply providing a method to convert to x,y data, but I don't  think it makes much difference. One suggestion I have is to clearly document the performance implications of using the  getitem/iterator methods, as this could throw people off quite easily.

**Comment by alexforencich** [comment]
Yeah, documenting that is a good idea.  I would remove it if it didn't completely break the existing functionality.  Even making this change will probably break some things, but I think it's worth the performance benefits.

**Comment by tobiasgehring** [comment]
just wondering: any thoughts on finalizing this? Can I be of assistance?
